feat: add deposit logic to Collateral

ramenforbreakfast opened the pull request:
This PR contains the `deposit` logic and associated tests. Some items that might be needed as context for the reviewer:
- Added a dev comment clarifying why we do not use the old fee logic that tries to prevent a user from skipping fees.
- I decided to keep just the `amountAfterFee` and the `fee` in the event, since the amount minted is already communicated via `Transfer('0x0'...` and the `amountBeforeFee` can be determined from these values.
- Instead of reverting upon `_amount > _fee` as specified in our tech doc (this is how we did it in the old contract), I did some thinking and realized there is no need for this since math is already underflow safe and there is no situation where this could happen except when the fee is 100%. A deposit of 0 would have been caught by this statement, but its easier just to revert if 0.

ramenforbreakfast commented:
Updated branch to reflect `Collateral` changes merged in recent PRs.

XavierEkkel commented on apps/smart-contracts/core/contracts/Collateral.sol:
```diff
@@ -49,7 +49,25 @@ contract Collateral is
   }
 
-  function deposit(uint256 _amount) external override {}
+  /**
+   * @dev Depending on how small the deposit fee is set, the fee might be
+   * zero. However, in a practical scenario, even a deposit fee as low as
+   * 0.1% would require a depositor to deposit less than $0.001 to not pay
+   * a fee. Such an attack would not be economically feasible and is not worth
+   * protecting against.
+   */
```
this is only true assuming the collateral is USDC, it may be economically feasible in other cases (or on other chains)
i think we should just have a check that _fee > 0 if depositFee > 0

XavierEkkel commented on apps/smart-contracts/core/contracts/Collateral.sol:
```diff
@@ -49,7 +49,25 @@ contract Collateral is
+   * protecting against.
+   */
+  function deposit(uint256 _amount) external override {
+    require(_amount > 0, "cannot deposit zero");
```
```suggestion
    require(_amount > 0, "amount = 0");
```

XavierEkkel commented on apps/smart-contracts/core/contracts/Collateral.sol:
```diff
@@ -49,7 +49,25 @@ contract Collateral is
+  function deposit(uint256 _amount) external override {
+    require(_amount > 0, "cannot deposit zero");
+    uint256 _fee = (_amount * _depositFee) / FEE_DENOMINATOR;
+    uint256 _amountAfterFee = _amount - _fee;
```
only declare variables right before they are needed
so this should be moved above the if statement

XavierEkkel commented on apps/smart-contracts/core/contracts/Collateral.sol:
```diff
@@ -49,7 +49,25 @@ contract Collateral is
+    if (address(_depositHook) != address(0)) {
+      _depositHook.hook(msg.sender, _amount, _amountAfterFee);
+    }
+    _mint(msg.sender, (_amountAfterFee * 1e18) / _baseTokenDenominator);
```
why do we need to scale by 1e18 when amount should already be scaled by its decimals?

ramenforbreakfast replied:
This has been clarified in discord and will resolve this

XavierEkkel commented on apps/smart-contracts/core/contracts/Collateral.sol:
```diff
@@ -49,7 +49,25 @@ contract Collateral is
+   * a fee. Such an attack would not be economically feasible and is not worth
+   * protecting against.
+   */
+  function deposit(uint256 _amount) external override {
```
add reentrancy guard
same for withdraw+managerWithdraw

XavierEkkel commented on apps/smart-contracts/core/contracts/interfaces/ICollateral.sol:
```diff
@@ -6,6 +6,12 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 interface ICollateral is IERC20Upgradeable {
+  event Deposit(
+    address indexed depositor,
+    uint256 indexed amount,
```
let's call this `amountAfterFee` to be clearer

XavierEkkel commented on apps/smart-contracts/core/contracts/interfaces/ICollateral.sol:
```diff
@@ -6,6 +6,12 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 interface ICollateral is IERC20Upgradeable {
+  event Deposit(
+    address indexed depositor,
+    uint256 indexed amount,
+    uint256 indexed fee
```
why are we indexing on amounts?

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -463,4 +495,149 @@ describe('=> Collateral', () => {
+      await collateral.connect(deployer).setDepositFee(TEST_DEPOSIT_FEE)
+      await collateral.connect(deployer).setDepositHook(depositHook.address)
+    })
+    it('reverts if depositing zero', async () => {
```
nit: keep concise
```suggestion
    it('reverts if deposit 0', async () => {
```

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -463,4 +495,149 @@ describe('=> Collateral', () => {
+      expect(await baseToken.balanceOf(treasury.address)).to.eq(treasuryBTBefore.add(fee))
+    })
+
+    it('mints equivalent Collateral to depositor based on base token decimals', async () => {
```
```suggestion
    it('mints to depositor adjusting for base token decimals', async () => {
```

XavierEkkel replied:
add assertions that the decimals are different for base vs. collateral

XavierEkkel replied:
add test for when base+collateral decimals are the same, and a test for when base decimals > collateral decimals

XavierEkkel replied:
add test for when deposit fee is set to 0

ramenforbreakfast commented:
Submitted various fixes and amendments:
- Changed setup to allow for dynamic base token setups, which will allow us to cleanly test base tokens with different decimal precision.
- Changed revert case to revert if calculated fee is 0 and fee rate > 0 as stated in the review. Otherwise, check if the amount = 0 (if amount = 0 when fee rate > 0 then it will be caught be the first condition).
- Made only `depositor` in the `Deposit` event indexed, changed `amount` to `amountAfterFee`
- Added testcases for different base token decimals, and tests for the new revert logic, and ensuring deposits can still happen when fee = 0.

XavierEkkel commented on apps/smart-contracts/core/contracts/Collateral.sol:
```diff
@@ -49,7 +51,22 @@ contract Collateral is
+      require(_amount > 0, "amount = 0");
+    }
+    _baseToken.transferFrom(msg.sender, address(this), _amount);
+    _baseToken.transfer(_treasury, _fee);
```
this is a logic error (and is why we shouldn't use `treasury`, because it could refer to either collateral treasury or prePO treasury - the two should be kept separate)

ramenforbreakfast commented:
Rebased and edited to use the new `manager` instead of `treasury` variable, removed fee transfers and instead approve fee funds to the deposit hook.

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -568,4 +601,237 @@ describe('=> Collateral', () => {
+      await collateral.connect(deployer).setDepositHook(depositHook.address)
+    })
+
+    it('reverts if deposit 0 and fee rate 0', async () => {
```
nit:
```suggestion
    it('reverts if deposit = 0 and deposit fee = 0%', async () => {
```

XavierEkkel replied:
similarly for other tests

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -568,4 +601,237 @@ describe('=> Collateral', () => {
+      expect(await collateral.getDepositFee()).to.be.gt(0)
+      const amountToDeposit = BigNumber.from(1)
+      // expect fee to be zero
+      expect(amountToDeposit.mul(await collateral.getDepositFee()).div(FEE_DENOMINATOR)).to.eq(0)
```
nit: assign the argument to a feeAmount variable first for readability, and then can also remove the comment above

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -568,4 +601,237 @@ describe('=> Collateral', () => {
+      await expect(collateral.connect(user1).deposit(amountToDeposit)).to.be.reverted
+    })
+
+    it('transfers amount to contract from depositor', async () => {
```
nit:
```suggestion
    it('transfers amount from depositor to contract', async () => {
```

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -568,4 +601,237 @@ describe('=> Collateral', () => {
+      expect(await collateral.balanceOf(user1.address)).to.eq(userCTBefore.add(expectedCT))
+    })
+
+    it('allows a deposit > 0 if fee rate 0', async () => {
```
nit:
```suggestion
    it('allows deposit > 0 if deposit fee = 0%', async () => {
```

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -568,4 +601,237 @@ describe('=> Collateral', () => {
+      const contractBTBefore = await baseToken.balanceOf(collateral.address)
+      const userCTBefore = await collateral.balanceOf(user1.address)
+      const amountToDeposit = userBTBefore
+      // expect fee to be zero
+      expect(amountToDeposit.mul(await collateral.getDepositFee()).div(FEE_DENOMINATOR)).to.eq(0)
```
nit: same comment here as before

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -568,4 +601,237 @@ describe('=> Collateral', () => {
+      expect(await baseToken.allowance(collateral.address, depositHook.address)).to.eq(secondFee)
+    })
+
+    it('mints to depositor adjusting for when decimals > base token decimals', async () => {
```
nit:
```suggestion
    it('mints decimal-adjusted amount to depositor if decimals > base token decimals', async () => {
```

XavierEkkel replied:
similarly for other tests

XavierEkkel reviewed (approved):
LGTM after nits

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -568,4 +601,237 @@ describe('=> Collateral', () => {
+
+      await collateral.connect(user1).deposit(secondAmountToDeposit)
+
+      expect(await baseToken.allowance(collateral.address, depositHook.address)).to.eq(secondFee)
```
add assertion that fee balance stays in the collateral contract
